src/libstore/worker-protocol-impl.hh

```diff
@@ -9,70 +9,50 @@
  */
 
 #include "worker-protocol.hh"
+#include "length-prefixed-protocol-helper.hh"
 
 namespace nix {
 
-template<typename T>
-std::vector<T> WorkerProto::Serialise<std::vector<T>>::read(const Store & store, WorkerProto::ReadConn conn)
-{
-    std::vector<T> resSet;
-    auto size = readNum<size_t>(conn.from);
-    while (size--) {
-        resSet.push_back(WorkerProto::Serialise<T>::read(store, conn));
-    }
-    return resSet;
-}
+/* protocol-agnostic templates */
 
-template<typename T>
-void WorkerProto::Serialise<std::vector<T>>::write(const Store & store, WorkerProto::WriteConn conn, const std::vector<T> & resSet)
-{
-    conn.to << resSet.size();
-    for (auto & key : resSet) {
-        WorkerProto::Serialise<T>::write(store, conn, key);
+#define WORKER_USE_LENGTH_PREFIX_SERIALISER(TEMPLATE, T) \
+    TEMPLATE T WorkerProto::Serialise< T >::read(const Store & store, WorkerProto::ReadConn conn) \
+    { \
+        return LengthPrefixedProtoHelper<WorkerProto, T >::read(store, conn); \
+    } \
+    TEMPLATE void WorkerProto::Serialise< T >::write(const Store & store, WorkerProto::WriteConn conn, const T & t) \
+    { \
+        LengthPrefixedProtoHelper<WorkerProto, T >::write(store, conn, t); \
     }
```

Comment:
Could we instead define the relevant instances as inheriting from `LenghtFixedProtoHelper` directly? Something like

```cpp
template <typename T>
struct WorkerProto::Serialise<std::set<T>> : LengthPrefixedProtocolHelper<WorkerProto, T> {};
```

Reply:
I tried that, but it didn't seem to be working (not sure why though). Moreover, it means `LengthPrefixedProtocolHelper` can't just be an implemnetation detail in the "impl" header, but must also be exposed in the main "interface" header.